pkg/jaegerproxy/proxy.go

```diff
@@ -0,0 +1,173 @@
+}
+
+func New(config ProxyConfig, logger hclog.Logger) (*Proxy, error) {
+	var opts []grpc.DialOption
+	var err error
+	if config.TLS {
+		if config.CaFile == "" {
+			return nil, fmt.Errorf("ca file is required with TLS")
+		}
+		creds, err := credentials.NewClientTLSFromFile(config.CaFile, config.ServerHostOverride)
+		if err != nil {
+			log.Fatalf("Failed to create TLS credentials %v", err)
+		}
+		opts = append(opts, grpc.WithTransportCredentials(creds))
+	} else {
+		opts = append(opts, grpc.WithInsecure())
+	}
+
+	if config.ConnectTimeout == 0 {
+		config.ConnectTimeout = DefaultTimeout
+	}
+
+	opts = append(opts, grpc.WithBlock())
+	ctx, cancel := context.WithTimeout(context.Background(), config.ConnectTimeout)
+	defer cancel()
+	conn, err := grpc.DialContext(ctx, config.ServerAddr, opts...)
+	if err != nil {
+		log.Fatalf("fail to dial: %v", err)
+	}
+
+	return &Proxy{
+		config:                  config,
+		logger:                  logger,
+		conn:                    conn,
+		spanReaderClient:        storage_v1.NewSpanReaderPluginClient(conn),
+		dependencyReaderClient:  storage_v1.NewDependenciesReaderPluginClient(conn),
+		capClient:               storage_v1.NewPluginCapabilitiesClient(conn),
+		archiveSpanReaderClient: storage_v1.NewArchiveSpanReaderPluginClient(conn),
+		archiveSpanWriterClient: storage_v1.NewArchiveSpanWriterPluginClient(conn),
+	}, nil
+}
+
+func (p *Proxy) Close() {
+	p.conn.Close()
+}
+
+func (p *Proxy) GetDependencies(ctx context.Context, r *storage_v1.GetDependenciesRequest) (*storage_v1.GetDependenciesResponse, error) {
+	return p.dependencyReaderClient.GetDependencies(ctx, r)
+}
+
+func (p *Proxy) WriteSpan(ctx context.Context, r *storage_v1.WriteSpanRequest) (*storage_v1.WriteSpanResponse, error) {
+	//TODO: figure out what to do here if we want the jaeger-all-in-one to work. Fine if using jaeger-query
+	return p.spanWriterClient.WriteSpan(ctx, r)
+}
+
+func (p *Proxy) GetTrace(r *storage_v1.GetTraceRequest, stream storage_v1.SpanReaderPlugin_GetTraceServer) error {
+	client, err := p.spanReaderClient.GetTrace(stream.Context(), r)
+	if err != nil {
+		return err
+	}
+	for {
+		m, err := client.Recv()
+		if err == io.EOF {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+		if err = stream.Send(m); err != nil {
+			return err
+		}
+	}
+}
+
+func (p *Proxy) GetServices(ctx context.Context, r *storage_v1.GetServicesRequest) (*storage_v1.GetServicesResponse, error) {
+	return p.spanReaderClient.GetServices(ctx, r)
+}
+
+func (p *Proxy) GetOperations(
+	ctx context.Context,
+	r *storage_v1.GetOperationsRequest,
+) (*storage_v1.GetOperationsResponse, error) {
+	return p.spanReaderClient.GetOperations(ctx, r)
+}
+
+func (p *Proxy) FindTraces(r *storage_v1.FindTracesRequest, stream storage_v1.SpanReaderPlugin_FindTracesServer) error {
+	client, err := p.spanReaderClient.FindTraces(stream.Context(), r)
+	if err != nil {
+		return err
+	}
+	for {
```

Comment:
Since streaming logic repeats in few places it's a good candidate for a function

Reply:
Good catch